### PR TITLE
Fix AddParser

### LINKED ISSUE #202: Improve parsing for add and schedule
Initially, I wanted to reject other formats other than
1) add a task name
2) add do something by (time)
3) add attend something from (time) to (time)
so things like
4) add read something to (time) will fail but the problem is natty reads number as time (so it does not help when we try to add back to the title string)
This fails:`add i have alot of things to do for 2103`

What we can do
1) acknowledge that it will be wrong. provide ways to overcome it e.g. don't parse what is enclosed in inverted commas etc
2) check for a) by or b) from+to otherwise c) add everything to the task string. No throwing incorrect command format.
3) Make it compulsory to use /by or -by instead

comment here or update user guide on implementation

### PR DESCRIPTION
**Pull request by burnflare**
Fixes #202

### CODE REVIEW

File: src/main/java/seedu/agendum/logic/parser/Parser.java [hunk 1]
```diff
@@ -170,12 +170,8 @@ private Command prepareAdd(String args) {
...
             } else if (dateTimeMap.containsKey(ARGS_FROM) && dateTimeMap.containsKey(ARGS_TO)) {
                 return new AddCommand(title, dateTimeMap.get(ARGS_FROM), dateTimeMap.get(ARGS_TO));
-            } else if (!dateTimeMap.containsKey(ARGS_FROM) && !dateTimeMap.containsKey(ARGS_TO)
-                    && !dateTimeMap.containsKey(ARGS_BY)) {
-                return new AddCommand(title);
-            } else {
-                return new IncorrectCommand(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
-                        AddCommand.MESSAGE_USAGE));
+            } else  {
+                return new AddCommand(args);
```
Comment (burnflare) [inline src/main/java/seedu/agendum/logic/parser/Parser.java]: @rachx, a quick fix I made up. Is this valid? Basically, if we don't fit the 2 criteria, just send the entire incoming string directly to `AddCommand(Str)`

**Comment by burnflare** [comment]
@rachx Updated with quotation support. Can you check?

**Comment by rachx** [comment]
Schedule should be fine right?

How about the following input

> add return "animal farm". it will only save "animal farm"
> @fanwgwg i think we might need to change our user guide examples and screenshots after our decision

I am thinking if we should save with the inverted commas
e.g. task name "copy assignment from cs2101 to cs2103". it makes the first and last word un-searchable.

Instead of using " ", can we use something that the user is unlikely to use and easy to type
e.g. add --drop by seven eleven -- (2 or 3 leading and trailing slashes but don't save to title). thoughts?

**Comment by rachx** [comment]
oh ya, you might need to change the add tests in logic manager test

**Comment by burnflare** [comment]
Schedule is fine.

Possible... we could change the delimiter to something else. @INCENDE?

**Comment by fanwgwg** [comment]
@rachx for safe, I'll just remove the quotation marks in the screenshots

**Comment by burnflare** [comment]
@fanwgwg Don't remove them yet. Let's come to a decision first so that you don't need to do double work

**Comment by fanwgwg** [comment]
okay that's fine!

**Comment by burnflare** [comment]
@rachx please verify

File: src/main/java/seedu/agendum/logic/parser/Parser.java [hunk 2]
```diff
@@ -188,18 +187,19 @@ private Command prepareAdd(String args) {
...
-            if (dateTimeMap.containsKey(ARGS_FROM)
-                    && dateTimeMap.containsKey(ARGS_TO)) {
+            if (!hasDeadlineKeyword && hasStartTimeKeyword && hasEndTimeKeyword) {
                 return new AddCommand(title, dateTimeMap.get(ARGS_FROM), dateTimeMap.get(ARGS_TO));
             }
 
-            if (!dateTimeMap.containsKey(ARGS_FROM)
-                    && !dateTimeMap.containsKey(ARGS_TO)
-                    && !dateTimeMap.containsKey(ARGS_BY)) {
+            if (!hasDeadlineKeyword && !hasStartTimeKeyword && !hasEndTimeKeyword) {
```
Comment (burnflare) [inline src/main/java/seedu/agendum/logic/parser/Parser.java:207]: Nice!

**Comment by rachx** [comment]
@burnflare @fanwgwg @INCENDE 
Everyone please verify.

I change the escape character from " " to ' ' only (easier to type and people can still use double inverted commas as normal, no need to redo screenshot)

so 
1) add 'copy notes from 2101 to 2103' will add the task _copy notes from 2101 to 2103_
2) natty will still parse add copy notes from 2101 to 2103 => task _copy notes_ with time
3) if there is no ' ' , it will expect
a) no by, from and to keywords with time behind
b) by keyword with time behind only
c) from and to keyword with time behind only
(this is for more consistent behavior)

Some small problems
1) add by 8pm. It will add the task name by 8pm instead of complaining incorrect format is that fine?
2) add return 'book' will be saved as book instead of _return 'book'_. just get people to use double inverted commas instead?

**Comment by rachx** [comment]
hmmm fix and decide on this and check that everything works after fixing gcal?

**Comment by burnflare** [comment]
I did some light testing and it looks good. @INCENDE Could you test this too?

**Comment by rachx** [comment]
how about add by 8pm?

**Comment by burnflare** [comment]
Thinking about it.... there's noting wrong with accepting `add by 8pm` 💯 

If user inputs `add by 8pm`, we expect there to be a title before `by 8pm`. But since we didn't see a title, we ignore the date time and insert the entire thing as a title instead. IMO, this behaviour is better than the opposite (create a blank event that's due at 8pm).

**Comment by rachx** [comment]
Okay. Initially, I was thinking of incorrect format since the name is missing. It recognises add from 4pm to 8pm as incorrect format
Lets wait for @INCENDE to merge

**Comment by burnflare** [comment]
Assigning to @rachx as I can't approve my own PR. Feel free to approve and merge when you're ready.

**Review by rachx — APPROVED** [review]

**Comment by INCENDE** [comment]
good to go

**Pull request merged**